### PR TITLE
Write the TSLint configuration file to the temp directory during build.

### PR DESCRIPTION
**Pull request by lijunle**
- The TSLint configuration file can be used for [VSCode TSLint plugin](https://marketplace.visualstudio.com/items?itemName=eg2.tslint).
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/Microsoft/web-build-tools/pull/129%23discussion_r105621243%22%2C%20%22https%3A//github.com/Microsoft/web-build-tools/pull/129%23discussion_r105621439%22%2C%20%22https%3A//github.com/Microsoft/web-build-tools/pull/129%23discussion_r105710804%22%2C%20%22https%3A//github.com/Microsoft/web-build-tools/pull/129%23discussion_r105712938%22%2C%20%22https%3A//github.com/Microsoft/web-build-tools/pull/129%23discussion_r105746332%22%2C%20%22https%3A//github.com/Microsoft/web-build-tools/pull/129%23discussion_r105746908%22%2C%20%22https%3A//github.com/Microsoft/web-build-tools/pull/129%23discussion_r105747588%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%20fe933c0899d20881f06508df30250a50fa8aa445%20common/changes/write-config_2017-03-12-14-57.json%205%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Microsoft/web-build-tools/pull/129%23discussion_r105746332%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%5C%22%2Aafter%20executing%5C%22%5Cr%5Cn%5Cr%5Cn--%20Remember%20that%20this%20string%20ends%20up%20in%20the%20changelog%22%2C%20%22created_at%22%3A%20%222017-03-13T19%3A10%3A11Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/17131271%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/nickpape-msft%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20common/changes/write-config_2017-03-12-14-57.json%3AL1-11%22%7D%2C%20%22Pull%20b769efe01e62b990905b5e52bdebbba60105edab%20gulp-core-build-typescript/src/TSLintTask.ts%2015%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Microsoft/web-build-tools/pull/129%23discussion_r105621243%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22You%20should%20be%20able%20to%20do%20this%20with%20Gulp.%22%2C%20%22created_at%22%3A%20%222017-03-13T09%3A46%3A33Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/5010588%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/iclanton%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20can%20do%20it%20as%20%2Agulp-like%2A%20pattern%20like%20%5Bthis%5D%28http%3A//stackoverflow.com/a/23398200/1436671%29.%20Do%20you%20think%20it%20is%20OK%3F%22%2C%20%22created_at%22%3A%20%222017-03-13T16%3A57%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1296500%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lijunle%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20gulp-core-build-typescript/src/TSLintTask.ts%3AL117-127%22%7D%2C%20%22Pull%20b769efe01e62b990905b5e52bdebbba60105edab%20gulp-core-build-typescript/src/TSLintTask.ts%2026%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Microsoft/web-build-tools/pull/129%23discussion_r105621439%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Is%20the%20%60tslint.json%60%20file%20location%20configurable%20in%20the%20TSLint%20plugin%3F%22%2C%20%22created_at%22%3A%20%222017-03-13T09%3A47%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/5010588%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/iclanton%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yes.%20There%20is%20an%20option%20-%20%60tslint.configFile%60%20to%20config%20the%20%60tslint.config%60%20file.%5Cr%5Cn%5Cr%5Cn%3E%20https%3A//marketplace.visualstudio.com/items%3FitemName%3Deg2.tslint%22%2C%20%22created_at%22%3A%20%222017-03-13T16%3A49%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/1296500%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lijunle%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20gulp-core-build-typescript/src/TSLintTask.ts%3AL195-212%22%7D%2C%20%22Pull%20fe933c0899d20881f06508df30250a50fa8aa445%20gulp-core-build-typescript/src/TSLintTask.ts%2015%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Microsoft/web-build-tools/pull/129%23discussion_r105746908%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22If%20you%20want%20to%20block%20behind%20this%20stream%2C%20I%20would%20advice%20using%20merge%20plugin%20to%20merge%20this%20and%20the%20below%20streams.%5Cr%5Cn%5Cr%5CnHowever%2C%20I%20still%20think%20using%20gulp%20to%20write%20a%20single%20file%20is%20complete%20overkill.%20fs.writeFileSync%28%29%20will%20be%2010%25%20the%20%23%20of%20lines.%22%2C%20%22created_at%22%3A%20%222017-03-13T19%3A12%3A48Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/17131271%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/nickpape-msft%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20gulp-core-build-typescript/src/TSLintTask.ts%3AL117-127%22%7D%2C%20%22Pull%20fe933c0899d20881f06508df30250a50fa8aa445%20gulp-core-build-typescript/src/TSLintTask.ts%2035%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/Microsoft/web-build-tools/pull/129%23discussion_r105747588%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Also%2C%20it%20looks%20like%20this%20is%20the%20wrong%20type%20of%20stream.%20Gulp/Vinyl%20is%20using%20a%20Stream%20of%20GulpUtil.Files%20objects.%20A%20better%20way%20to%20accomplish%20what%20this%20function%20is%20doing%20would%20be%20to%20adopt%20some%20code%20from%20%60FileUtils.ts%60%20in%20%60%40microsoft/sp-build-core-tasks%60.%20IMO%2C%20we%20should%20migrate%20some%20of%20those%20utilities%20into%20directly%20into%20GCB.%22%2C%20%22created_at%22%3A%20%222017-03-13T19%3A15%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars3.githubusercontent.com/u/17131271%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/nickpape-msft%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20gulp-core-build-typescript/src/TSLintTask.ts%3AL195-212%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [x] <a href='#crh-comment-Pull b769efe01e62b990905b5e52bdebbba60105edab gulp-core-build-typescript/src/TSLintTask.ts 15'></a> &nbsp;<b><a href='https://github.com/Microsoft/web-build-tools/pull/129#discussion_r105621243'>File: gulp-core-build-typescript/src/TSLintTask.ts:L117-127</a></b>
- <a href='https://github.com/iclanton'></a> You should be able to do this with Gulp.
- <a href='https://github.com/lijunle'></a> I can do it as *gulp-like* pattern like [this](http://stackoverflow.com/a/23398200/1436671). Do you think it is OK?
- [x] <a href='#crh-comment-Pull b769efe01e62b990905b5e52bdebbba60105edab gulp-core-build-typescript/src/TSLintTask.ts 26'></a> &nbsp;<b><a href='https://github.com/Microsoft/web-build-tools/pull/129#discussion_r105621439'>File: gulp-core-build-typescript/src/TSLintTask.ts:L195-212</a></b>
- <a href='https://github.com/iclanton'></a> Is the `tslint.json` file location configurable in the TSLint plugin?
- <a href='https://github.com/lijunle'></a> Yes. There is an option - `tslint.configFile` to config the `tslint.config` file.
> https://marketplace.visualstudio.com/items?itemName=eg2.tslint
- [x] <a href='#crh-comment-Pull fe933c0899d20881f06508df30250a50fa8aa445 common/changes/write-config_2017-03-12-14-57.json 5'></a> &nbsp;<b><a href='https://github.com/Microsoft/web-build-tools/pull/129#discussion_r105746332'>File: common/changes/write-config_2017-03-12-14-57.json:L1-11</a></b>
- <a href='https://github.com/nickpape-msft'></a> "*after executing"
-- Remember that this string ends up in the changelog
- [x] <a href='#crh-comment-Pull fe933c0899d20881f06508df30250a50fa8aa445 gulp-core-build-typescript/src/TSLintTask.ts 15'></a> &nbsp;<b><a href='https://github.com/Microsoft/web-build-tools/pull/129#discussion_r105746908'>File: gulp-core-build-typescript/src/TSLintTask.ts:L117-127</a></b>
- <a href='https://github.com/nickpape-msft'></a> If you want to block behind this stream, I would advice using merge plugin to merge this and the below streams.
However, I still think using gulp to write a single file is complete overkill. fs.writeFileSync() will be 10% the # of lines.
- [x] <a href='#crh-comment-Pull fe933c0899d20881f06508df30250a50fa8aa445 gulp-core-build-typescript/src/TSLintTask.ts 35'></a> &nbsp;<b><a href='https://github.com/Microsoft/web-build-tools/pull/129#discussion_r105747588'>File: gulp-core-build-typescript/src/TSLintTask.ts:L195-212</a></b>
- <a href='https://github.com/nickpape-msft'></a> Also, it looks like this is the wrong type of stream. Gulp/Vinyl is using a Stream of GulpUtil.Files objects. A better way to accomplish what this function is doing would be to adopt some code from `FileUtils.ts` in `@microsoft/sp-build-core-tasks`. IMO, we should migrate some of those utilities into directly into GCB.

<a href='https://www.codereviewhub.com/Microsoft/web-build-tools/pull/129?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/Microsoft/web-build-tools/pull/129?approve=1'></a>&nbsp;<a href='https://github.com/Microsoft/web-build-tools/pull/129'></a>
<a href='#crh-end'></a>

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/web-build-tools/129)

### CODE REVIEW

File: gulp-core-build-typescript/src/TSLintTask.ts [hunk 1]
```diff
@@ -115,6 +117,10 @@ export class TSLintTask extends GulpTask<ITSLintTaskConfig> {
     const taskScope: TSLintTask = this;
 
     const activeLintRules: any = taskScope._loadLintRules(); // tslint:disable-line:no-any
+
+    // @todo This stream is not piped to returned value. Find a way to block the execute flow.
+    this._writeLintRules(activeLintRules);
```
Comment (iclanton) [inline gulp-core-build-typescript/src/TSLintTask.ts]: You should be able to do this with Gulp.

File: gulp-core-build-typescript/src/TSLintTask.ts [hunk 2]
```diff
@@ -189,4 +195,17 @@ export class TSLintTask extends GulpTask<ITSLintTaskConfig> {
       (this.taskConfig.useDefaultConfigAsBase ? this._defaultLintRules : {}),
       this.taskConfig.lintConfig || {});
   }
+
+  private _writeLintRules(lintRules: any): fs.WriteStream { // tslint:disable-line:no-any
+    const tslintFilePath: string = path.resolve(this.buildConfig.rootPath, this.buildConfig.tempFolder, 'tslint.json');
```
Comment (iclanton) [inline gulp-core-build-typescript/src/TSLintTask.ts]: Is the `tslint.json` file location configurable in the TSLint plugin?

Comment (lijunle) [inline gulp-core-build-typescript/src/TSLintTask.ts]: Yes. There is a option - `tslint.configFile` to config the `tslint.config` file.

> https://marketplace.visualstudio.com/items?itemName=eg2.tslint

File: gulp-core-build-typescript/src/TSLintTask.ts (hunk 1 above)
Comment (lijunle) [inline gulp-core-build-typescript/src/TSLintTask.ts]: I can do it as *gulp-like* pattern like [this](http://stackoverflow.com/a/23398200/1436671). Do you think it is OK?

File: common/changes/write-config_2017-03-12-14-57.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/gulp-core-build-typescript",
+      "comment": "Write the TSLint configuration file when execute.",
```
Comment (nickpape-msft) [inline common/changes/write-config_2017-03-12-14-57.json]: *after executing

File: gulp-core-build-typescript/src/TSLintTask.ts (hunk 1 above)
Comment (nickpape-msft) [inline gulp-core-build-typescript/src/TSLintTask.ts]: If you want to block behind this stream, I would advice using merge plugin to merge this and the below streams.

However, I still think using gulp to write a single file is complete overkill. fs.writeFileSync() will be 10% the # of lines.

File: gulp-core-build-typescript/src/TSLintTask.ts [hunk 4]
```diff
@@ -189,4 +195,17 @@ export class TSLintTask extends GulpTask<ITSLintTaskConfig> {
...
+    const tslintFileStream: fs.WriteStream = fs.createWriteStream(tslintFilePath);
+
+    const readable: Readable = new Readable();
+    const pipe: fs.WriteStream = readable.pipe(tslintFileStream);
+
+    readable.push(JSON.stringify(lintRules, undefined, 2));
+    readable.push(null); // tslint:disable-line:no-null-keyword
+
+    return pipe;
```
Comment (nickpape-msft) [inline gulp-core-build-typescript/src/TSLintTask.ts]: Also, it looks like this is the wrong type of stream. Gulp/Vinyl is using a Stream of GulpUtil.Files objects. A better way to accomplish what this function is doing would be to adopt [this](https://onedrive.visualstudio.com/SPPPlat/_git/sp-client?path=%2Fspfx-tools%2Fsp-build-core-tasks%2Fsrc%2Futilities%2FFileUtils.ts&version=GBmaster&_a=contents&line=4&lineStyle=plain&lineEnd=27&lineStartColumn=1&lineEndColumn=2). IMO, we should migrate some of those utilities into directly into GCB.

File: gulp-core-build-typescript/src/TSLintTask.ts (hunk 1 above)
Comment (lijunle) [inline gulp-core-build-typescript/src/TSLintTask.ts]: Cool. I could leverage the `fs.writeFileSync` API for simplicity.

File: gulp-core-build-typescript/src/TSLintTask.ts (hunk 4 above)
Comment (lijunle) [inline gulp-core-build-typescript/src/TSLintTask.ts]: This is not adopting the Vinyl file model. It is only node.js file stream.

Comment (lijunle) [inline gulp-core-build-typescript/src/TSLintTask.ts]: This is not adopting the Vinyl file model. It is only node.js file stream.

**Comment by nickpape-msft** [comment]
Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/web-build-tools/129)*

**Pull request merged**